src/puppetlabs/trapperkeeper/internal.clj

```diff
@@ -176,6 +176,29 @@
         required    []]
     (first (ks/cli! cli-args specs required))))
 
+(def exit-request-schema
+  "A process exit request like
+  {:status 7
+   :messages [[\"something for stderr\n\" *err*]]
+              [\"something for stdout\n\" *out*]]
+              [\"something else for stderr\n\" *err*]]"
+  {:status schema/Int
+   :messages [[(schema/one schema/Str "message")
+               (schema/one java.io.Writer "stream")]]})
+
+(defn exit-exception? [ex]
+  (and (instance? ExceptionInfo ex)
+       (not (schema/check {(schema/optional-key :puppetlabs.trapperkeeper.core/exit)
+                           exit-request-schema}
+                          (ex-data ex)))))
+
+(defn shutdown-reason-for-ex
+  [exception]
+  (if (exit-exception? exception)
+    (merge {:cause :requested}
+           (select-keys (ex-data exception) [:puppetlabs.trapperkeeper.core/exit]))
```

Comment:
Why `select-keys` instead of including all the exception data?
Also, based on your documentation, it looks like `:puppetlabs.trapperkeeper.core/exit` is not a top-level key, it's the value for the `:kind` key. so not sure what the intention was here.

Reply:
If I recall correctly, it's just to be sure that the exception only contributes the key that it's supposed to, e.g. can't clobber other top-level keys, etc.

And I think the idea is that `:kind` is a non-namespaced key that tells you what type of ex-data map you have.  Everything else in the map can be type-specific.  The ':kind` value is namespaced to avoid collisions across namespaces and even across different projects.

We could namespace `:kind`, as say `:puppetlabs/kind`, but I believe there's some precedence at puppet (and I vaguely recall maybe elsewhere) to use `:kind` globally.  TK itself is already using `:kind` for some exit-related code.